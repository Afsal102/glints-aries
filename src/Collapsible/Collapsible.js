/* @flow */

import React from 'react';
import createReactContext, { type Context } from 'create-react-context';
import {
  CollapsibleContainer,
  CollapsibleHeadStyle,
  CollapsibleBodyStyle,
} from '../Style/CollapsibleStyle';
import { Icon } from '../Icon';

type isOpen = true | false;
type children = React$Node;
const CollapsibleContext: Context<isOpen> = createReactContext(false);


const Collapsible = (props: Props) => {
  const {
    children,
    isOpen,
    style,
    title,
    ...defaultProps
  } = props;

  return (
    <CollapsibleContainer style={style} {...defaultProps}>
      <CollapsibleContext.Provider value={isOpen}>
        {children}
      </CollapsibleContext.Provider>
    </CollapsibleContainer>
  );
};

export const CollapsibleHead = (props: HeadProps) => {
  return (
    <CollapsibleContext.Consumer>
      {(isOpenContext) => {
        return (
          <CollapsibleHeadStyle>
            <span>
              {
                props.children
              }
            </span>
            {isOpenContext && <Icon name="arrow-up" size="12" color="#000000" />}
            {!isOpenContext && <Icon name="arrow-down" size="12" color="#000000" />}
          </CollapsibleHeadStyle>
        )
      }}
    </CollapsibleContext.Consumer>
  );
}

export const CollapsibleBody = (props: BodyProps) => {
  return (
    <CollapsibleContext.Consumer>
      {(isOpenContext) => {
        if (isOpenContext) {
          return (
            <CollapsibleBodyStyle>
              {
                props.children
              }
            </CollapsibleBodyStyle>
          );
        }

        return null;
      }}
    </CollapsibleContext.Consumer>
  );
};

type Props = {
  children?: children,
  style?: Object,
  isOpen?: isOpen,
  title: String,
}

type HeadProps = {
  children?: children,
  style?: Object,
}

type BodyProps = {
  children?: children,
  style?: Object,
}

Collapsible.defaultProps = {
  isOpen: false,
};

export default Collapsible;
