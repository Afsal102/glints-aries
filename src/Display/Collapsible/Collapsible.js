import React, { Component } from 'react';
import Icon from '../../General/Icon';

import {
  CollapsibleContainer,
  CollapsibleContent,
  CollapsibleHeader,
  CollapsibleBody,
} from '../../Style/Display/CollapsibleStyle';

class Collapsible extends Component <Props, State> {
  state = {
    isOpen: true,
  }

  handleOpen = () => {
    const { isOpen } = this.state;
    this.setState({ isOpen: !isOpen });
  }

  shouldComponentUpdate(nextProps, nextState) {
    const { isOpen } = this.state;
    if (isOpen !== nextState.isOpen) {
      // re-render
      return true;
    }

    // nothing changed. avoid unnecessary re-render
    return false;
  }

  render() {
    const {
      label,
      children,
      className,
      ...defaultProps
    } = this.props;

    const { isOpen } = this.state;

    return (
      <CollapsibleContainer
        className="collapsible"
        tabIndex="0"
        onClick={this.handleOpen}
        {...defaultProps}
      >
        <CollapsibleContent tabIndex="-1">
          <CollapsibleHeader
            className="head"
            role="tab"
            aria-expanded={isOpen}
          >
            {label}
            <Choose>
              <When condition={isOpen === false}>
                <Icon name="arrow-down" color="#000000" />
              </When>
              <Otherwise>
                <Icon name="arrow-up" color="#000000" />
              </Otherwise>
            </Choose>
          </CollapsibleHeader>
          {isOpen && (
            <CollapsibleBody className="body">
              {children}
            </CollapsibleBody>
          )}
        </CollapsibleContent>
      </CollapsibleContainer>
    );
  }
}

type Props = {
  children: React$Node,
  className: string,
  label: string,
};

type State = {
  isOpen: boolean,
}

export default Collapsible;
