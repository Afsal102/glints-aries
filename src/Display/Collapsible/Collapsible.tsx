import * as React from 'react';
import classNames from 'classnames';
import { ArrowDownIcon } from '../../General/Icon/components';

import {
  CollapsibleContainer,
  CollapsibleContent,
  CollapsibleHeader,
  CollapsibleBody,
} from '../../Style/Display/CollapsibleStyle';

class Collapsible extends React.Component<Props, State> {
  state = {
    isOpen: true,
  };

  handleOpen = () => {
    const { isOpen } = this.state;
    this.setState({ isOpen: !isOpen });
  };

  componentDidMount() {
    const { isOpen } = this.props;

    if (isOpen === false && isOpen !== undefined) {
      this.setState({
        isOpen,
      });
    }
  }

  render() {
    const { label, children, className, ...defaultProps } = this.props;

    const { isOpen } = this.state;

    return (
      <CollapsibleContainer
        className={classNames('aries-collapsible', className)}
        tabIndex={0}
        onClick={this.handleOpen}
        data-testid="collapsible-container"
        {...defaultProps}
      >
        <CollapsibleContent tabIndex={-1}>
          <CollapsibleHeader
            className="collapsible-title"
            role="tab"
            aria-expanded={isOpen}
            isOpen={isOpen}
          >
            {label}
            <ArrowDownIcon color="#000000" />
          </CollapsibleHeader>
          {isOpen && (
            <CollapsibleBody
              className="collapsible-content"
              data-testid="collapsible-content"
              onClick={e => e.stopPropagation()}
            >
              {children}
            </CollapsibleBody>
          )}
        </CollapsibleContent>
      </CollapsibleContainer>
    );
  }
}

interface Props
  extends React.ComponentPropsWithoutRef<typeof CollapsibleContainer> {
  label: React.ReactNode;
  isOpen?: boolean;
}

interface State {
  isOpen: boolean;
}

export default Collapsible;
