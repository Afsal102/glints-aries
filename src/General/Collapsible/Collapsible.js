import React, { Component } from 'react';
import Icon from '../Icon';

import { Container, Header, Body } from '../../Style/CollapsibleStyle';

class Collapsible extends Component <Props, State> {
  state = {
    isOpen: false,
  }

  handleOpen = () => {
    const { isOpen } = this.state;
    this.setState({ isOpen: !isOpen });
  }

  componentDidMount() {
    const { collapsing } = this.props;

    if (collapsing) {
      this.setState({
        isOpen: false,
      });
    }
  }

  render() {
    const {
      label,
      children,
      className,
      ...defaultProps
    } = this.props;

    const { isOpen } = this.state;

    return (
      <Container className="collapsible" {...defaultProps} onClick={this.handleOpen}>
        <Header className="head">
          {label}
          <Choose>
            <When condition={isOpen === false}>
              <Icon name="arrow-down" color="#000000" />
            </When>
            <Otherwise>
              <Icon name="arrow-up" color="#000000" />
            </Otherwise>
          </Choose>
        </Header>
        {isOpen && (
          <Body className="body">
            {children}
          </Body>
        )}
      </Container>
    );
  }
}

type Props = {
  children: React$Node,
  className: string,
  label: string,
};

type State = {
  isOpen: boolean,
}

export default Collapsible;
