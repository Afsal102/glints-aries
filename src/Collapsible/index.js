// @flow

import Collapsible, { CollapsibleHead, CollapsibleBody } from './Collapsible';

Collapsible.Head = CollapsibleHead;
Collapsible.Body = CollapsibleBody;

export default Collapsible;
