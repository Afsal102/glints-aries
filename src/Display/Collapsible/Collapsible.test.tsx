import * as React from 'react';

import * as renderer from 'react-test-renderer';
import '@testing-library/jest-dom/extend-expect';
import { fireEvent, render } from '@testing-library/react';

import Collapsible from './Collapsible';

const defaultCollapsible = <Collapsible label="label">children</Collapsible>;

const closedCollapsible = (
  <Collapsible label="label" isOpen={false}>
    children
  </Collapsible>
);

const openCollapsible = (
  <Collapsible label="label" isOpen={true}>
    children
  </Collapsible>
);

const setupCollapsible = () => {
  const { getByRole, getByTestId } = render(defaultCollapsible);
  return {
    header: getByRole('tab'),
    container: getByTestId('collapsible-container'),
    body: getByTestId('collapsible-content'),
  };
};

function createNodeMock(element: any) {
  if (element.type === 'div') {
    return {
      contains: jest.fn(),
    };
  }
  return null;
}

describe('<Collapsible>', () => {
  it('should render as expected', () => {
    const options = { createNodeMock };
    const closedCollapsibleSnapshot = renderer
      .create(closedCollapsible, options)
      .toJSON();
    expect(closedCollapsibleSnapshot).toMatchSnapshot();

    const openCollapsibleSnapshot = renderer
      .create(openCollapsible, options)
      .toJSON();
    expect(openCollapsibleSnapshot).toMatchSnapshot();
  });

  it('should be open by default', () => {
    const { header } = setupCollapsible();
    expect(header).toHaveAttribute('aria-expanded', 'true');
  });

  it('should close on click', () => {
    const { header, container } = setupCollapsible();
    fireEvent.click(container);
    expect(header).toHaveAttribute('aria-expanded', 'false');
  });

  it('should open on second click', () => {
    const { header, container } = setupCollapsible();
    fireEvent.click(container);
    fireEvent.click(container);
    expect(header).toHaveAttribute('aria-expanded', 'true');
  });

  it('should not propagate click events from CollapsibleBody', () => {
    const { body } = setupCollapsible();
    const mockEventListener = jest.fn();
    window.addEventListener('click', mockEventListener);
    fireEvent.click(body);
    expect(mockEventListener).not.toHaveBeenCalled();
  });
});
