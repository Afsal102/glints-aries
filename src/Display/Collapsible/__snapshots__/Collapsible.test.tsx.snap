// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Collapsible> should render as expected 1`] = `
<div
  className="aries-collapsible CollapsibleStyle__CollapsibleContainer-sc-16akyym-1 ctRcDY"
  data-testid="collapsible-container"
  onClick={[Function]}
  tabIndex={0}
>
  <div
    className="CollapsibleStyle__CollapsibleContent-sc-16akyym-0 lddXIm"
    tabIndex={-1}
  >
    <div
      aria-expanded={false}
      className="collapsible-title CollapsibleStyle__CollapsibleHeader-sc-16akyym-2 cCEtWo"
      role="tab"
    >
      label
      <svg
        className="IconStyle__VerticalCenteredSvg-sc-19ci3uf-0 dmBUgt"
        data-testid="icon-svg"
        fill="#000000"
        height="1em"
        viewBox="0 0 100 100"
        width="1em"
      >
        <polygon
          points="11.7984725 19 49.9935275 57.3950291 88.1885825 19 99.987055 30.7984725 49.9935275 80.792 0 30.7984725"
        />
      </svg>
    </div>
  </div>
</div>
`;

exports[`<Collapsible> should render as expected 2`] = `
<div
  className="aries-collapsible CollapsibleStyle__CollapsibleContainer-sc-16akyym-1 ctRcDY"
  data-testid="collapsible-container"
  onClick={[Function]}
  tabIndex={0}
>
  <div
    className="CollapsibleStyle__CollapsibleContent-sc-16akyym-0 lddXIm"
    tabIndex={-1}
  >
    <div
      aria-expanded={true}
      className="collapsible-title CollapsibleStyle__CollapsibleHeader-sc-16akyym-2 gqXdzD"
      role="tab"
    >
      label
      <svg
        className="IconStyle__VerticalCenteredSvg-sc-19ci3uf-0 dmBUgt"
        data-testid="icon-svg"
        fill="#000000"
        height="1em"
        viewBox="0 0 100 100"
        width="1em"
      >
        <polygon
          points="11.7984725 19 49.9935275 57.3950291 88.1885825 19 99.987055 30.7984725 49.9935275 80.792 0 30.7984725"
        />
      </svg>
    </div>
    <div
      className="collapsible-content CollapsibleStyle__CollapsibleBody-sc-16akyym-3 ldgfso"
      data-testid="collapsible-content"
      onClick={[Function]}
    >
      children
    </div>
  </div>
</div>
`;
